### ISSUE TITLE
DASH Live: DefaultDashChunkSource incorrectly assumes that a period of known duration declares all segments

### ISSUE THREAD

**Issue by AnjelloNordija** [google/ExoPlayer#8538]
We have a leanback launcher which uses ExoPlayer. When pausing the live stream, we shift from the live stream to a timeshifted stream and then pause it. When the user presses resume, the time shifted stream resumes playback. This works fine with all our client video streams but with one particular client.

With this particular client, after the timeshifted stream is resumed, the player freezes at the end of the first audio/video segment. However this does not happen if the time shifted stream is resumed within a few seconds.

There are a few differences on where the content is originated from and how the manifest is built.

- For LIVE TV we use a dynamic manifest with 32 s of segments. Content is sourced from a circular buffer.

- For time-shift (pause live) the manifest is requested from the circular buffer with start and stop times. When the program reaches the stop time the manifest changes from dynamic to static. Until that point in time it is dynamic.

- For catch-up, start-over and N-PVR it depends if the program has finished recording and is moved to the SAN or not when the request hist the request router. If on SAN it is always gets a static manifest. If not finished it follows the same rules as time-shift.

I have tried to extract a timeshifted stream url and playing it on the ExoPlayer demo app. And I can reproduce this same freezing with the same error on logs (D/AudioTrack: stop() called with xxxxxx frames delivered) every time. I have tried on ExoPlayer demo versions 2.11.4 and 2.12.3, therefore think that this is a player/manifest incompatibility.

The same manifest is used for start over without any issues. 

I have mailed the log files to dev.exoplayer@gmail.com

Thanks in Advance,
Anjello

**Comment by icbaker** [comment]
Please can you provide a manifest + instructions that we can use to reproduce the problem?

**Comment by DrewsWaoo** [comment]
We have a lab setup where I can give you access to a few test streams so you can reproduce the problem anytime you want.

I'll send the details to dev.exoplayer@gmail.com.

BR
Torben Drews

**Comment by icbaker** [comment]
Thanks that sounds useful - let us know when you've sent the details over.

**Comment by AnjelloNordija** [comment]
Once you receive the test URL from Torben, you need to add startTime and stopTime parameters to it. The startTime needs to be within the last 30 minutes, and in the Zulu format. So for instance if the current time is 20210215T102500Z, the URL could look like:

https://xxxxxxxxxx/manifest.mpd?token=xxxxxxxxxxxxxxxxxxxxxxxx&expires=1615967284&startTime=20210215T100500Z&stopTime=20210215T100600Z"

Thanks,
Anjello

**Comment by icbaker** [comment]
I'm afraid I can't provide the info requested by @DrewsWaoo over email - the stream needs to be accessible from any device to ensure everyone on the team is able to reproduce it as we all have different areas of expertise. 

If you're unable to host the stream in a way we can access and that reproduces the problem then I'm afraid there's not much we can do to help.

**Comment by AnjelloNordija** [comment]
Hi icbaker,

Torben has replied on mail with stream URLs with the security tokens disabled.

**Comment by icbaker** [comment]
Thanks - I'm able to reproduce the problem as follows:
1. Adjust the 'circular buffer' URL provided over email to ensure that the `startTime` parameter is less than 30mins in the past and `stopTime` is in the future.
1. Play the URL

The video & audio will freeze after a short period. The playback position continues to count up.

I noticed that when I start playback, the ExoPlayer demo app UI shows the 'position' as something like XX:XX / YY:YY. As playback advances both of these numbers increase (i.e. the 'current' position and 'duration' of the content). The audio & video freeze when the 'current' position reaches the initial 'duration'. I guess this is just a rephrasing of your original post:
> the player freezes at the end of the first audio/video segment.

@ojw28 can you take a look from a DASH perspective?

**Comment by ojw28** [comment]
These manifests are unusual because they define a period with a known duration (the period duration is known due to the `mediaPresentationDuration` attribute on the `MPD` element), but to which segments are still being appended.

ExoPlayer doesn't appear to handle this case correctly. This is because `DefaultDashChunkSource` assumes that a period is complete if its duration is known:

https://github.com/google/ExoPlayer/blob/r2.13.1/library/dash/src/main/java/com/google/android/exoplayer2/source/dash/DefaultDashChunkSource.java#L340

It assumes that no further segments will be appended to such a period:

https://github.com/google/ExoPlayer/blob/r2.13.1/library/dash/src/main/java/com/google/android/exoplayer2/source/dash/DefaultDashChunkSource.java#L366

This means that if the player loads the last segment in the timeline of whatever manifest it currently has, then it wont try and load anything more, which is the point at which playback freezes.

I think this is an ExoPlayer issue, however if you don't _need_ to include `mediaPresentationDuration` until the manifest changes to static, them omitting it until that point will be the easiest/fastest way for you to resolve this issue.

**Comment by AnjelloNordija** [comment]
@owj28,

Thanks for the feedback. Will this issue be fixed on a future release of ExoPlayer? If so, how far down the line will that be?

Best Regards,
Anjello

**Comment by DrewsWaoo** [comment]
I don't know if it will be easy to omit the `mediaPresentationDuration` as the streaming platform is a commercial product and if I understand your answer correct the manifest comply the the DASH standard. Then it is not a bug with SLA - but instead a feature request to be able to remove a tag in order to make an open source player to work. It is not impossible - but I don't know what will be fastest. But I'll give it a try.

I still don't understand the choice/confusing in the player... I thought it was the `type="dynamic|static"` that should signal the player whether to keep fetching the manifest or not.

BR
Torben Drews

**Comment by ojw28** [comment]
> Will this issue be fixed on a future release of ExoPlayer? If so, how far down the line will that be?

Yes, although I don't have a timeframe at the moment. The correct fix requires a bit of further thought.

> I still don't understand the choice/confusing in the player... I thought it was the type="dynamic|static" that should signal the player whether to keep fetching the manifest or not.

The problem isn't that the player stops refreshing the manifest. Manifest refreshes are still happening. The problem is an implementation detail where some logic deeper in the player thinks that no further segments will be appended to the period. `type="dynamic|static"` isn't sufficient to determine whether this is the case, because in multi-period DASH you can have a period for which all content is available, and also still have `type="dynamic"`. One option would be to check whether the period duration is known *and* the last declared segment extends to the end of the period, but we've definitely seen examples of DASH content where the segments end short of the period duration, so this is likely to break more cases than we'd be fixing. As above, working out a good fix, that doesn't break any other cases, will require a bit of careful thought.
